hibernate-core/src/main/java/org/hibernate/collection/internal/PersistentBag.java

```diff
@@ -131,21 +134,46 @@ public void beforeInitialize(CollectionPersister persister, int anticipatedSize)
 	}
 
 	@Override
+	@SuppressWarnings( "unchecked" )
 	public boolean equalsSnapshot(CollectionPersister persister) throws HibernateException {
 		final Type elementType = persister.getElementType();
-		final List sn = (List) getSnapshot();
+		final List<Object> sn = (List<Object>) getSnapshot();
 		if ( sn.size() != bag.size() ) {
 			return false;
 		}
-		for ( Object elt : bag ) {
-			final boolean unequal = countOccurrences( elt, bag, elementType ) != countOccurrences( elt, sn, elementType );
-			if ( unequal ) {
+
+		// HHH-11032 - Group objects by Type.getHashCode() to reduce the complexity of the search
+		final Map<Integer, List<Object>> hashToInstancesBag = groupByEqualityHash( bag, elementType );
+		final Map<Integer, List<Object>> HashToInstancesSn = groupByEqualityHash( sn, elementType );
+		if ( hashToInstancesBag.size() != HashToInstancesSn.size() ) {
+			return false;
+		}
+
+		for ( Map.Entry<Integer, List<Object>> hashToInstancesBagEntry : hashToInstancesBag.entrySet() ) {
+			final int hash = hashToInstancesBagEntry.getKey();
+			final List<Object> instancesBag = hashToInstancesBagEntry.getValue();
+			final List<Object> instancesSn = HashToInstancesSn.getOrDefault( hash, Collections.emptyList() );
```

Comment:
rather than `getOrDefault` you could just return `false` when the get returns null.

Reply:
```getOrDefault``` was in the @barreiro update to the original commit in ```getItemsWithSameHash```. I thought about changing it, but I kept it, for readability, so I wouldn't need to add a check for ```null``` in the following ```if``` statement. 

I suppose checking for ```null``` would be less expensive than ```size```.

@Sanne, could you respond for future reference? I'm not against changing it.

Reply:
No strong opinion but I'd change it:

> I suppose checking for null would be less expensive than size.

+1 and it saves invoking `Collections.emptyList()` as  well.

Reply:
OK, I'll change it.